api/sys_raft.go

```diff
@@ -84,28 +86,60 @@ func (ac *AutopilotConfig) UnmarshalJSON(b []byte) error {
 
 // AutopilotState represents the response of the raft autopilot state API
 type AutopilotState struct {
-	Healthy          bool                        `mapstructure:"healthy"`
-	FailureTolerance int                         `mapstructure:"failure_tolerance"`
-	Servers          map[string]*AutopilotServer `mapstructure:"servers"`
-	Leader           string                      `mapstructure:"leader"`
-	Voters           []string                    `mapstructure:"voters"`
-	NonVoters        []string                    `mapstructure:"non_voters"`
+	Healthy                    bool                        `mapstructure:"healthy"`
+	FailureTolerance           int                         `mapstructure:"failure_tolerance"`
+	Servers                    map[string]*AutopilotServer `mapstructure:"servers" deep:"-"`
```

Comment:
I'm guessing the `deep` annotation is to tell deep.Equal to ignore this field?  If so I don't really like this, I feel it's liable to be surprising to test writers.  What motivated it?

Reply:
Yes, that's what it's for. The motivation for it was cleaning up the ent tests, which is a bit of a long story. Previous to this change, the repeated pattern in the ent tests was to poll the state endpoint in a loop, and do a bunch of `require()` calls from the testify library to verify the contents of the state endpoint. That approach was problematic though, because autopilot converges on its end state slowly, so any `require()` calls that didn't match immediately aborted the test.

Then I discovered that the testify library had `assert()` calls which were like require, but didn't abort the test - just returned a bool. So I switched to those, but it was still flaky. Then I tried a long cascading series of if statements, but that was gross. I'll spare you all the other things I tried, but what I finally settled on for all the cases except for the leadership transfer was constructing a state object showing what I expected the end result to be, polling the state endpoint, and calling `deep.Equal()` on the two state objects until they matched.

This worked beautifully and had the added bonus of showing very plainly what the expected state should be, right in the test. `Servers` is simply a field that isn't relevant to the ent tests, which is why I ignored it here.

I realize this might be sub-optimal, but imo, the benefits it conveys today far outweighs the possibility of future test writers getting tripped up by it.

Reply:
What the cause of the mis-match, ordering?

Reply:
I'm sorry, I'm not following. Mis-match?

Reply:
Well, you're excluding a field from the deep.Equal comparison, and I'm wondering why in the absence of that exclusion, deep.Equal is saying the state doesn't match?

Reply:
Also note that https://github.com/google/go-cmp is a more flexible deep.Equal equivalent that wouldn't prevent impact other tests trying to do comparisons on this.  I'm just worried that someone is going to waste hours debugging a test, wondering why things aren't behaving as they expect, because of this little-known feature.

Reply:
Excluding that field allows me to skip the laborious process of specifying the state of all the servers, which I don't care about for any of these tests.

I was not aware of `go-cmp`. I do wonder if now is the right time to consider rewriting these tests to use a new equality library.

I understand your concern, but I also think the odds of someone writing brand new tests in the future that leverage the autopilot state is somewhat low. Perhaps we can table this discussion for now and take an action item to revisit this when time is not in such short supply.

Reply:
I think the change is smaller than you think, here's an example:

```
func deepEqualTokenEntries(t *testing.T, a *logical.TokenEntry, b *logical.TokenEntry) {
	if diff := cmp.Diff(a, b, cmpopts.IgnoreFields(logical.TokenEntry{}, "ExternalID")); diff != "" {
		t.Fatalf("bad diff in token entries: %s", diff)
	}
}
```

Reply:
If you like I could attempt the change.